### PR TITLE
i #313 Added Notebook for Syntax Extraction

### PR DESCRIPTION
**Pull request by daomcgill**
- Documented current syntax extraction functions
- Overview on syntax extraction and XPath
- Placeholder for syntax.yml config file

### CODE REVIEW

**Review by carlosparadis — CHANGES_REQUESTED** [review]
These should be double `../../` because when we run the code block, it assumes it is located on vignettes. I believe this also needs to be fixed on the mail PR.

File: conf/syntax.yml
Comment (carlosparadis) [inline conf/syntax.yml:1]: This conf name does not make sense, the config files are per project. Syntax is not a project!

File: vignettes/syntax_extractor.Rmd [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+  require(data.table)
+  require(yaml)
+  require(stringi)
+  require(XML)
+  require(gt)
+```
+
+# Introduction
+
+In open-source projects, code is often spread across many files and organized in various ways. Understanding this code at a glance can be difficult, especially when projects grow large and complex. This is where syntax extraction comes in.
```
Comment (carlosparadis) [inline vignettes/syntax_extractor.Rmd:29]: I would remove this. The motivation fits better to why someone would need Kaiaulu than the feature you are working here!

File: vignettes/syntax_extractor.Rmd [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+
+The Syntax Extractor is a tool that helps us dive into the structure of source code by using an annotated format. It works by converting code into a structured XML format using srcML, making it easier to analyze and understand the underlying architecture of the project. By extracting key pieces of information such as class names, functions, or even comments (like documentation), we can generate meaningful data that helps with things like semantic analysis, code comprehension, and more.
```
Comment (carlosparadis) [inline vignettes/syntax_extractor.Rmd:31]: You should reword this to not give the allusion this Notebook is a tool of its own. For example, I would expect this to be on a README of a standalone tool with the feature you are working on, rather than the explanation of a new capability in Kaiaulu.  

Rather you could start by something like: `This notebook explains how to convert` and then tie to what you have on `code into a structured XML format using the tool srcML`. 

I would also remove the `making it easier to analyze and understand the underlying architecture of the project`. Rather, the motivation is to `selectively extract terms of a source file, so it may be represented for other analysis, such as`. and connect to `semantic analysis, code comprehension, etc.`

File: vignettes/syntax_extractor.Rmd [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+Imagine a codebase as a collection of books in a library. Without a catalog, finding the right book would be nearly impossible. Syntax extraction is like building a catalog for a codebase. It allows us to quickly pinpoint where classes, methods, and important comments are located, and organize this data in a structured way.
```
Comment (carlosparadis) [inline vignettes/syntax_extractor.Rmd:33]: Analogies are very good, but I'm not sure if we want to "sell the fish" here as syntax extraction being a way to catalogue source code to find classes. Instead, I'd like you to try another analogy, that builds on your original motivation of semantic analysis and code comprehension. 

Food for thought for the analogy: Why would we want to do semantic analysis and code comprehension assessments from extracting the file?

File: vignettes/syntax_extractor.Rmd [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+For Kaiaulu, this is particularly useful as we move toward deeper analysis, such as applying machine learning techniques like word embeddings to the extracted syntax. This allows us to not only understand how the code is written but also how the pieces relate to each other on a semantic level.
```
Comment (carlosparadis) [inline vignettes/syntax_extractor.Rmd:35]: This comes closer to the spirit of the analogy: Why would we want a machine to understand semantically how two classes relate? How would this be helpful for a developer?

File: vignettes/syntax_extractor.Rmd [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+
+In short, syntax extraction gives us a way to “see the big picture” without having to manually dig through every line of code.
```
Comment (carlosparadis) [inline vignettes/syntax_extractor.Rmd:37]: I would like you to emphasize a bit more the fact that, representing a file through different methods, can lead to different representations, and consequently understanding by a computer that is comparing files using this information!

For example, consider that we could a) show all the file content to a computer or b) only the names of classes, variables and documentation, which capture semantic meaning tied to the code, or c) commits and issues discussion associated to the file (just like we are doing now), which capture others discussion surrounding the file. 

You could give a computer different representations (a, b, c) of a given file, and ask the computer to compare to another file (also represented in either a, b, c). Consider a computer would be having a different understanding of what two files are related depending on the representation of your choice:

It may say using criteria c) that the files are alike, because generally people are discussing similar features on its commits and issues, but if using a) the way the files were implemented being substantially different, they could be different.

File: vignettes/syntax_extractor.Rmd [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+
+## How the Syntax Extractor Works
+
+At its core, the syntax extractor relies on [srcML](https://www.srcml.org/), which is a tool that converts source code into an XML representation. This XML structure gives us a detailed breakdown of the code: classes, functions, variables, and comments are annotated with specific XML tags, allowing us to query and extract what we need.
```
Comment (carlosparadis) [inline vignettes/syntax_extractor.Rmd:41]: I would defer this to after the project configuration file, so you can explain how it works as you call functions bit by bit, interspersed with examples and runnable code.

File: vignettes/syntax_extractor.Rmd [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+
+In this notebook, we’ll walk through the process of setting up the syntax extractor, running the extraction process, and then querying the annotated code for useful information like class names, namespaces, and documentation.
+
+# Project Configuration File
+
+In a project, source code is spread across multiple repositories and folders. To use Kaiaulu’s syntax extraction functions, you need to configure the system with details about where the source code is stored and where the tools (like srcML) are located.
+
+Kaiaulu uses a project configuration file format to specify the paths and settings necessary for syntax extraction. This allows you to manage different projects and their associated codebases efficiently.
```
Comment (carlosparadis) [inline vignettes/syntax_extractor.Rmd:49]: I would replace this with more precise motivation. For example, `manage different projects codebases efficiently` does not tell me much. 

For those two sentences, you want to focus on the fact that a) Kaiaulu project configuration files contain information about a **project** that is of interest, b) where you want to store the input and output to the tool, and c) what configuration parameters you want out of the tool (i.e. the tool may have its own arguments, the config is where you store them). The reason is reproducibility. If I don't know what you used as input, and the arguments, then I can't re-run and expect the same result.

File: vignettes/syntax_extractor.Rmd [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+tool:
+  srcml:
+    srcml_path: /usr/local/bin/srcml
+    srcml_output_path: ../../analysis/maven/srcml_output.xml
+
+analysis:
+  topics:
+    - class
+    - method
+    - documentation
```
Comment (carlosparadis) [inline vignettes/syntax_extractor.Rmd:74]: this probably will need revisiting later, it is legacy of another thing we were doing with srcML. We can go over this on call

File: vignettes/text_gof_showcase.Rmd
Comment (carlosparadis) [inline vignettes/text_gof_showcase.Rmd:1]: should undo the `../` changes

File: vignettes/syntax_extractor.Rmd [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+)
+
+# Display the extracted comments
+function_comments %>%
+  gt()
+```
+
+This query will return a list of comments associated with each function in the code.
+
+### XPath Cheat Sheet
```
Comment (carlosparadis) [inline vignettes/syntax_extractor.Rmd:221]: We should make an actual Cheatsheet for Kaiaulu for this later on.

**Review by carlosparadis — CHANGES_REQUESTED** [review]
I have not done the full review yet, but I hope this helps. Is my understanding correct that the documentation extractor is still pending? 

We can do another full pass on the notebook at that point after you try it out.

File: conf/syntax.yml
Comment (daomcgill) [reply conf/syntax.yml:1]: Will change the name. I am a little bit confused on the ../../ since I thought we were supposed to set out Global Options -> R Markdown -> Evaluate chunks in directory -> Project?

**Comment by daomcgill** [comment]
@carlosparadis I made the changes we discussed. I also added functions for extracting function names (opt. params) and variable names (opt. types). I do need to do a more thorough manual check to make sure the results are accurate, but it appears to be working so far. 
Should I continue adding functions--maybe imports or some way of determining inheritance structure (I would need to think about this some more)--or move on and come back once the other pieces start to come together?

**Comment by carlosparadis** [comment]
Working on this PR now, please hold on doing commits.

**Comment by carlosparadis** [comment]
@daomcgill i suggest you fix this PR first with the get and the other notes, assuming this is the one that came before the exec PR. Then you can merge locally this branch to the exec PR, to avoid having to do it twice (just be careful you dont merge the one that came after to the PR before or it will mix everything up).

**Comment by daomcgill** [comment]
@carlosparadis okay, I will work on this now

**Comment by daomcgill** [comment]
@carlosparadis fixed hardcoded paths and reverted workflow. Going to move to 314 now

**Comment by carlosparadis** [comment]
@daomcgill could you confirm the notebook display the parsed tables so the users can see what each function outputs as i did with parse_mbox on your mail notebook?

**Comment by daomcgill** [comment]
@carlosparadis every query that I run displays a table. Should I allow every query to be evaluated then? I think that also means I have to change it so the annotation is evaluated?

**Comment by daomcgill** [comment]
@carlosparadis I think I answered my own question. Notebook now displays tables.

**Comment by carlosparadis** [comment]
@daomcgill you sure did! It is very much in line with the changes I did on mail and that you see on itm0, there is no better way to showcase your work on the docs than having a table showing on the intermediate steps. Please send me on drive the files needed to run the code (I imagine just the raw data suffice, the rest I can extract from your functions). Thank you!

**Comment by daomcgill** [comment]
@carlosparadis I just used the cloned maven repository but, yes, I am adding it to the drive right now (folder is just named maven). 

I think I have made it through all the requested changes, in both branches. I have also updated the fasttext notebook, and it appears to be running with the new config setup.

**Comment by carlosparadis** [comment]
Thank you!!
